### PR TITLE
Python net specification

### PR DESCRIPTION
**Pull request by longjon**
Protobuf is powerful, but is not a general-purpose language, and can be cumbersome in its current use for writing down complicated chains of functions, i.e., nets. See previous discussion at (nonexhaustively) #1290 and #1169.

Python is a general-purpose language with clean syntax and an intimate relationship with Caffe. This PR allows net specification as (domain-specific) Python code (and consequently, effortless programmatic net construction).

Currently this is just a proof-of-concept sketch, which lightly sprinkles Python magic over the `caffe_pb2` protobuf interface.

LeNet can be written as:
```python
from caffe import layers as L
from caffe import params as P

data = L.data(batch_size=64, backend=P.data.LMDB, source='examples/mnist/mnist_train_lmdb')
conv1 = L.convolution(data, kernel_size=5, num_output=20)
pool1 = L.pooling(conv1, kernel_size=2, stride=2, pool=P.pooling.MAX)
conv2 = L.convolution(pool1, kernel_size=5, num_output=50)
pool2 = L.pooling(conv2, kernel_size=2, stride=2, pool=P.pooling.MAX)
ip1 = L.relu(L.inner_product(pool2, num_output=500))
prob = L.softmax(L.inner_product(ip1, num_output=10))
```

And can be instantiated via this hackiness, to be cleaned up later:
```python
with open('tmp.prototxt', 'w') as f:
    print >>f, prob.to_proto()
net = caffe.Net('tmp.prototxt')
```

Things to think about:
* Perhaps there should be a way to construct `Net`s directly from `Layer`s, without going through protobuf.
* The magic in use here means syntax doesn't really gets checked until the `Net` is created.
* This will need to be rebased on #1694 and more significantly #1474.
* It would be nicer if the functions constructed here simply _were_ `Net`s rather than having to make `Net`s out of them.
* There is no way to control blob names right now, but there are ways to make that happen...

### CODE REVIEW

**Comment by netheril96** [comment]
> Perhaps there should be a way to construct Nets directly from Layers, without going through protobuf.

This is because the constructor of `net` unnecessarily involves I/O. It should instead take a `NetParameter` as argument, instead of (or in addition to) a filename as string.

**Comment by jeffdonahue** [comment]
This looks nifty.  What happens with multiple tops? (e.g. in your example, can you do `data, label = L.data(...)`?)

> the constructor of net unnecessarily involves I/O. It should instead take a NetParameter as argument, instead of (or in addition to) a filename as string.

I agree, there should definitely be a way to construct a pycaffe Net by passing a caffe_pb2.NetParameter thing (if you have one) rather than writing it to a file -- PRs welcome (or I'll probably try to do this at some point if nobody else does).

**Comment by longjon** [comment]
Re: multiple tops, I had in mind your syntax @jeffdonahue, but I don't think I implemented it here yet.

Yes, there should be a way to construct `Net`s from `NetParameter`s, although that was not what I had in mind with the comment above, but rather a way to construct `Net`s without involving protobuf at all. In this sketch, one constructs some special Python class which specifies a net and then turns it into a `Net`. Why doesn't one just construct a `Net` to begin with?

**Comment by longjon** [comment]
Just pushed an update to this; still a bit hacky, but now more feature complete:
* rebased on string types
* supports layers with multiple tops (with an explicit `ntop` kwarg)
* supports in-place layers (`in_place` kwarg)
* supports output of nets with multiple disconnected tops
* supports `transform_param`, `loss_param`, and other non-layer-specific params
* the automatic names have been replaced with an explicit name dict, which, combined with an (arguably atrocious) hack allows the actual Python variable names to be used as blob/layer names

Here's one way to generate train and test for an AlexNet variant (striped down for unrelated reasons) with the current code:
```python
from caffe import layers as L, params as P, to_proto

# helper functions for common structures

def conv_relu(bottom, ks, nout, stride=1, pad=0, group=1):
    conv = L.Convolution(bottom, kernel_size=ks, stride=stride,
                                num_output=nout, pad=pad, group=group)
    return conv, L.ReLU(conv, in_place=True)

def fc_relu(bottom, nout):
    fc = L.InnerProduct(bottom, num_output=nout)
    return fc, L.ReLU(fc, in_place=True)

def max_pool(bottom, ks, stride=1):
    return L.Pooling(bottom, pool=P.Pooling.MAX, kernel_size=ks, stride=stride)

def alexnet(lmdb, batch_size=256, include_acc=False):
    data, label = L.Data(source=lmdb, backend=P.Data.LMDB, batch_size=256, ntop=2,
        transform_param=dict(crop_size=227, mean_value=[104, 117, 123], mirror=True))

    # the net itself
    conv1, relu1 = conv_relu(data, 11, 96, stride=4)
    pool1 = max_pool(relu1, 3, stride=2)
    conv2, relu2 = conv_relu(pool1, 5, 256, pad=2, group=2)
    pool2 = max_pool(relu2, 3, stride=2)
    conv3, relu3 = conv_relu(pool2, 3, 384, pad=1)
    conv4, relu4 = conv_relu(relu3, 3, 384, pad=1, group=2)
    conv5, relu5 = conv_relu(relu4, 3, 256, pad=1, group=2)
    pool5 = max_pool(relu5, 3, stride=2)
    fc6, relu6 = fc_relu(pool5, 4096)
    fc7, relu7 = fc_relu(relu6, 4096)
    fc8 = L.InnerProduct(relu7, num_output=1000)
    loss = L.SoftmaxLoss(fc8, label)

    if include_acc:
        acc = L.Accuracy(fc8, label)
        return to_proto((loss, acc), {v: k for k, v in locals().iteritems()})
    else:
        return to_proto(loss, {v: k for k, v in locals().iteritems()})

with open('train.prototxt', 'w') as f:
    print >>f, alexnet('/path/to/caffe-train-lmdb')

with open('test.prototxt', 'w') as f:
    print >>f, alexnet('/path/to/caffe-val-lmdb', batch_size=50, include_acc=True)
```

I'll be dogfooding this along presently.

File: python/caffe/layers.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+def uncamel(s):
+    """Convert CamelCase to underscore_case."""
+
+    return re.sub('(?!^)([A-Z]+)', r'_\1', s).lower()
+
+def assign_proto(proto, name, val):
+    if isinstance(val, list):
+        getattr(proto, name).extend(val)
+    else:
```
Comment (sontran) [inline python/caffe/layers.py]: To generate weight filler & bias filler, I would add the following here 
elif isinstance(val, dict):
        new_proto = getattr(proto, name)
        for k, v in val.iteritems():
            assign_proto(new_proto, k, v)

File: python/caffe/layers.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
+from collections import OrderedDict
+import re
+
+from .proto import caffe_pb2
+from google import protobuf
+
+def uncamel(s):
+    """Convert CamelCase to underscore_case."""
+
+    return re.sub('(?!^)([A-Z]+)', r'_\1', s).lower()
```
Comment (sontran) [inline python/caffe/layers.py]: Doesn't yet work for "LRN"

**Comment by longjon** [comment]
Closing to reopen against master.

**Comment by kevin-li** [comment]
Is there a way to modify existing network from within python? Or is there a way to edit prototxt by coding instead of manully changing them? 

My experiment requires freezing/training different layers, how can I do this through python code? 

Thank you.
